File: core/src/domain/offer.rs
//! Offer entity
//!
//! Ported from legacy/domain/entity/entity.go:86-106

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

use super::{ItemStatus, UrgencyLevel};

/// Represents a medication supply offer
/// Ported from Go: Offer struct (entity.go:86-106)
#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct Offer {
    pub id: String,
    pub raw_message_id: String,
    pub source_phone: String,
    pub source_name: String,
    pub source_group: String,
    pub group_name: String,
    pub medication: String,
    pub medication_raw: String,
    pub quantity: f64,
    pub unit: Option<String>,
    pub price: f64,
    pub currency: Option<String>,
    pub expiry_date: Option<DateTime<Utc>>,
    pub batch_number: Option<String>,
    pub notes: Option<String>,
    pub raw_message: String,
    pub status: ItemStatus,
    pub content_embedding: Option<Vec<f32>>,
    /// Urgency flag (seller urgency to sell)
    #[serde(default)]
    pub urgent: bool,
    /// Granular urgency level
    #[serde(default)]
    pub urgency_level: UrgencyLevel,
    /// Expiry info from AI extraction (YYYY-MM or description)
    pub expiry_info: Option<String>,
    /// AI extraction confidence score
    #[serde(default)]
    pub ai_confidence: f64,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

impl Default for Offer {
    fn default() -> Self {
        let now = Utc::now();
        Self {
            id: uuid::Uuid::new_v4().to_string(),
            raw_message_id: String::new(),
            source_phone: String::new(),
            source_name: String::new(),
            source_group: String::new(),
            group_name: String::new(),
            medication: String::new(),
            medication_raw: String::new(),
            quantity: 0.0,
            unit: None,
            price: 0.0,
            currency: None,
            expiry_date: None,
            batch_number: None,
            notes: None,
            raw_message: String::new(),
            status: ItemStatus::Active,
            content_embedding: None,
            urgent: false,
            urgency_level: UrgencyLevel::Normal,
            expiry_info: None,
            ai_confidence: 0.0,
            created_at: now,
            updated_at: now,
        }
    }
}

File: core-v2/packages/domain/src/entities/offer.ts
/**
 * Offer entity schema and types
 */
import { Schema } from "@effect/schema";
import { OfferId, GroupJid, SenderJid, RawMessageId } from "../ids";
import { ItemStatus } from "../enums";

export const Offer = Schema.Struct({
  id: OfferId,
  medication: Schema.String,
  medicationRaw: Schema.NullOr(Schema.String),
  quantity: Schema.NullOr(Schema.String),
  price: Schema.NullOr(Schema.String),
  expiry: Schema.NullOr(Schema.Date),
  status: ItemStatus,
  groupJid: Schema.NullOr(GroupJid),
  senderJid: Schema.NullOr(SenderJid),
  rawMessageId: Schema.NullOr(RawMessageId),
  aiConfidence: Schema.NullOr(Schema.Number),
  embedding: Schema.NullOr(Schema.Array(Schema.Number)),
  confirmedMatchCount: Schema.Number,
  createdAt: Schema.Date,
  updatedAt: Schema.Date,
});

export type Offer = Schema.Schema.Type<typeof Offer>;

export const CreateOfferInput = Schema.Struct({
  medication: Schema.String,
  medicationRaw: Schema.optional(Schema.String),
  quantity: Schema.optional(Schema.String),
  price: Schema.optional(Schema.String),
  expiry: Schema.optional(Schema.Date),
  groupJid: Schema.optional(GroupJid),
  senderJid: Schema.optional(SenderJid),
  rawMessageId: Schema.optional(RawMessageId),
  aiConfidence: Schema.optional(Schema.Number),
  embedding: Schema.optional(Schema.Array(Schema.Number)),
});

export type CreateOfferInput = Schema.Schema.Type<typeof CreateOfferInput>;

export const UpdateOfferInput = Schema.Struct({
  medication: Schema.optional(Schema.String),
  medicationRaw: Schema.optional(Schema.String),
  quantity: Schema.optional(Schema.String),
  price: Schema.optional(Schema.String),
  expiry: Schema.optional(Schema.Date),
  status: Schema.optional(ItemStatus),
  aiConfidence: Schema.optional(Schema.Number),
  embedding: Schema.optional(Schema.Array(Schema.Number)),
});

export type UpdateOfferInput = Schema.Schema.Type<typeof UpdateOfferInput>;
